### PR TITLE
Fix structlog JSON serialization crash on non-serializable objects

### LINKED ISSUE #62201: structlog configuration fails with dict input
### Apache Airflow version

3.1.7

### If "Other Airflow 3 version" selected, which one?

_No response_

### What happened?

structlog error was raised when a dict is passed as an argument to a logging message.

### What you think should happen instead?

It should be able to print the string representation of the dict passed as an argument

### How to reproduce

Run this file. You can see that it succeeds and behaves as expected if you remove the `import airflow`

```py
import logging
import airflow

logging.getLogger(__name__).warning('Info message %s', {'a': 10})
```

### Operating System

Debian GNU/Linux 12 (bookworm)

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

_No response_

### Anything else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #62472: Logging dictionary object in task fails (structlog/stdlib misconfiguration)
### Apache Airflow version

main (development)

### If "Other Airflow 3 version" selected, which one?

_No response_

### What happened?

Task that does:
d = {'a': 3}
logging.info('with format %s', d)

produces error log messages (but task succeeds)

### What you think should happen instead?

_No response_

### How to reproduce

1. Run breeze
2. Deploy DAG
```
from airflow.models.dag import DAG
from airflow.sdk import task
from datetime import datetime

import logging

@task
def print_dict():
    d = {'a': 3}
    print(d) # <- ok
    logging.info(d) # <- ok
    logging.info(f'with f string {d}') # <- ok
    logging.info('with format %s', d) # <- ERROR

with DAG(
    dag_id="print_dict_dag",
    start_date=datetime(2023, 1, 1),
    schedule=None,
    catchup=False,
) as dag:
    t0 = print_dict()
```
3. Observe

Image

### Operating System

Linux

### Versions of Apache Airflow Providers

_No response_

### Deployment

Other

### Deployment details

Breeze

### Anything else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by Vamsi-klu**
## Summary

- Wrap `enc_hook` in `json_dumps()` with a `safe_default()` fallback that catches `TypeError` and falls back to `str()`
- Prevents the logging pipeline from crashing when log event dicts contain objects that `msgspec` cannot serialize
- Defensive fallback only engages when normal serialization paths fail
- Hoisted closure into a module-level factory `_make_safe_enc_hook()` for cleaner code and testability

Co-contributors: @codingrealitylabs @girlcoder-gaming

closes: #62472
closes: #62201

## Test plan

- [x] `test_json_non_serializable_object` — verifies non-serializable objects fall back to `str()` instead of crashing
- [x] `test_json_custom_object_uses_repr` — regression test confirming normal custom objects still serialize via `repr()`
- [x] `test_safe_enc_hook_with_none_default` — unit test for the `default=None` edge case
- [x] All 51 structlog tests pass locally
- [x] Ruff lint and format pass

---

##### Was generative AI tooling used to co-author this PR?

- [X] Yes — Claude Code (Claude Opus 4.6)

Generated-by: Claude Code (Claude Opus 4.6) following [the guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#gen-ai-assisted-contributions)

### CODE REVIEW

**Comment by Vamsi-klu** [comment]
cc @ashb @amoghrajesh @potiuk — Would appreciate your review. This adds a `str()` fallback in the structlog JSON serializer to prevent logging pipeline crashes on non-serializable objects.

**Comment by potiuk** [comment]
You need to (as always) add test to that change - including covering the case that crashed before from the linked issues.

File: shared/logging/src/airflow_shared/logging/structlog.py [hunk 1]
```diff
@@ -316,7 +316,17 @@ def json_dumps(msg, default):
                 "event": msg.pop("event"),
                 **msg,
             }
-            return msgspec.json.encode(msg, enc_hook=default)
+
+            def safe_default(obj):
+                """Fall back to str() if the default enc_hook fails or is None."""
+                if default is not None:
```
Comment (potiuk) [inline shared/logging/src/airflow_shared/logging/structlog.py]: **Missing tests.** This is a bug fix that changes serialization behavior, but no tests were added. Please add tests to `shared/logging/tests/logging/test_structlog.py` covering:

1. A log event with a non-serializable object (e.g., a custom class instance) in the event dict doesn't crash and falls back to `str()`.
2. A log event with standard serializable objects still works correctly (existing `test_json` covers this, but a targeted test for the `safe_default` path would be valuable).
3. The case where `default` is `None` (if that's possible in practice).

Example:
```python
def test_json_non_serializable_object(structlog_config):
    class Unserializable:
        def __str__(self):
            return "<Unserializable>"

    with structlog_config(json_output=True) as bio:
        logger = structlog.get_logger("my.logger")
        logger.info("Hello", obj=Unserializable())

    written = json.load(bio)
    assert written["obj"] == "<Unserializable>"
```

Comment (potiuk) [inline shared/logging/src/airflow_shared/logging/structlog.py]: **Minor performance:** `safe_default` is a new closure allocated on every call to `json_dumps` (i.e., every JSON log line). Consider hoisting it out of the function body:

```python
def _make_safe_default(default):
    def safe_default(obj):
        if default is not None:
            try:
                return default(obj)
            except TypeError:
                pass
        return str(obj)
    return safe_default

def json_dumps(msg, default):
    msg = {
        "timestamp": msg.pop("timestamp"),
        "level": msg.pop("level"),
        "event": msg.pop("event"),
        **msg,
    }
    return msgspec.json.encode(msg, enc_hook=_make_safe_default(default))
```

Or, if `default` is always the same callable, cache it once. That said, this is a minor concern — closure creation is cheap in Python. Up to you whether to address it.

**Review by potiuk — CHANGES_REQUESTED** [review]
The fix is reasonable in principle — a logging pipeline should never crash due to a non-serializable object in the log context. However, the PR is missing test coverage for the new behavior, and the `safe_default` closure is unnecessarily recreated on every log line.

### What's good
- Defensive approach: only catches `TypeError`, not broad `Exception`
- Correctly checks `default is not None` before calling it
- `str()` returns a type that `msgspec` can always serialize

### What needs attention
1. **No tests added** — this is a bug fix that should have at least one test proving non-serializable objects don't crash the JSON logger, and one test confirming normal serialization still works through the same path.
2. **Minor performance concern** — `safe_default` is re-created as a closure on every `json_dumps` call (i.e., every JSON log line). It could be defined once.

File: shared/logging/src/airflow_shared/logging/structlog.py (hunk 1 above)
Comment (Vamsi-klu) [reply shared/logging/src/airflow_shared/logging/structlog.py]: Addressed in 1bd14c45f2. Added three tests:

1. **`test_json_non_serializable_object`** — an object whose `__structlog__()` raises `TypeError` now falls back to `str()` instead of crashing. This is the actual crash path: `_json_fallback_handler` only catches `AttributeError`, so `TypeError` from `__structlog__()` propagates through msgspec and crashes the logger.

2. **`test_json_custom_object_uses_repr`** — regression test confirming normal custom objects (without `__structlog__`) still serialize via `repr()` through the standard `_json_fallback_handler` path.

3. **`test_safe_enc_hook_with_none_default`** — unit test for the `default=None` edge case, testing `_make_safe_enc_hook` directly.

Comment (Vamsi-klu) [reply shared/logging/src/airflow_shared/logging/structlog.py]: Addressed in 1bd14c45f2. Hoisted the closure into a module-level factory function `_make_safe_enc_hook(default)` — follows the same pattern as the existing `_make_airflow_structlogger` in the same file.

The `json_dumps` body is now a clean one-liner:
```python
return msgspec.json.encode(msg, enc_hook=_make_safe_enc_hook(default))
```

The factory is at module level (not inside the `@cache`d `structlog_processors`) so it's also directly importable for unit testing.

**Comment by Vamsi-klu** [comment]
@potiuk can you please review the change and give me feedback? Thanks!

**Comment by eladkal** [comment]
static checks are failing. can you fix it?

**Comment by Vamsi-klu** [comment]
Hi @eladkal, thank you for flagging this! The static check failures have been addressed in commit 4352e27. The two issues were:

1. **Ruff docstring formatting** — the docstring in `_ensure_exec_plugin_cache_dirs()` needed the summary on a separate line from the opening `"""`.
2. **Ruff import sorting** — imports in `test_kubernetes.py` needed to be reordered alphabetically.

The branch has also been rebased onto the latest `main`. CI should be green on this run.

**Comment by Vamsi-klu** [comment]
Hi @jscheffl, thank you for the careful review and for catching this. You're absolutely right — the KubernetesHook changes (`_ensure_exec_plugin_cache_dirs` for issue #60943) were unrelated to the structlog fix and should not have been part of this PR. They were not hallucinated code, but they were mistakenly committed to this branch instead of their own.

I've removed those commits from this PR. The branch now contains only the two structlog-related commits:

1. **`6a6d698`** — The core fix: wrap `enc_hook` with a `safe_default` fallback that catches `TypeError` and falls back to `str()`
2. **`7bd90e2`** — Review feedback: hoist the closure into a module-level factory `_make_safe_enc_hook()`, and add three targeted tests

The KubernetesHook fix will be submitted as a separate PR. Apologies for the noise, and thank you again for flagging it.

File: shared/logging/src/airflow_shared/logging/structlog.py [hunk 2]
```diff
@@ -224,6 +224,20 @@ def respect_stdlib_disable(logger: Any, method_name: Any, event_dict: EventDict)
...
 
 
+def _make_safe_enc_hook(default):
+    """Wrap an enc_hook so that serialization failures fall back to ``str()``."""
+
+    def safe_enc_hook(obj):
+        if default is not None:
+            try:
+                return default(obj)
+            except TypeError:
```
Comment (jscheffl) [inline shared/logging/src/airflow_shared/logging/structlog.py]: In DockerOperator (fixed in https://github.com/apache/airflow/pull/62632) we saw:
```
UnicodeEncodeError: 'utf-8' codec can't encode character '\udce2' in position 76: surrogates not allowed
```
...which sub-classes from `ValueError`. So Unicode problems are not catched by this. Can you generalize this?

Comment (Vamsi-klu) [reply shared/logging/src/airflow_shared/logging/structlog.py]: Hey @jscheffl, thank you for pointing this out. I took a closer look and you're right, the surrogate issue needs handling at two levels.

The enc_hook now catches both TypeError and ValueError, so encoding errors during object conversion are covered. But the trickier part was that surrogate characters in string values bypass the enc_hook entirely since msgspec handles strings natively and only fails when it tries to encode them to UTF-8. So I also added a fallback around the msgspec.json.encode() call itself that replaces surrogates with the Unicode replacement character and retries.

I've added two tests for this. One covers the ValueError path in the enc_hook directly, and the other reproduces the surrogate scenario you described from the DockerOperator, making sure it doesn't crash and the rest of the log entry comes through cleanly.

All 53 structlog tests pass locally and ruff is clean. Let me know if this looks good to you.

**Review by jscheffl — APPROVED** [review]
For me this looks good now.

@potiuk would be good to have a second pair of eyes on this.

**Comment by Vamsi-klu** [comment]
@potiuk can i get you final check on this PR?

**Review by potiuk — APPROVED** [review]
